src/redisearch_rs/result_processor/src/search_result.rs

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+use enumflags2::{BitFlags, bitflags};
+use inverted_index::RSIndexResult;
+use std::ptr::{self, NonNull};
+
+use crate::bindings::DocumentMetadata;
+
+#[bitflags]
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum SearchResultFlag {
+    ExpiredDoc = 1,
+}
+
+pub type SearchResultFlags = enumflags2::BitFlags<SearchResultFlag>;
+
+// /*
+//  * SearchResult - the object all the processing chain is working on.
+//  * It has the indexResult which is what the index scan brought - scores, vectors, flags, etc,
+//  * and a list of fields loaded by the chain
+//  */
+#[derive(Clone, Debug)]
+pub struct SearchResult<'a> {
+    doc_id: ffi::t_docId,
+    // not all results have score - TBD
+    score: f64,
+
+    /// Raw pointer to the [`ffi::RSScoreExplain`].
+    ///
+    /// # Safety
+    ///
+    /// The pointer must be a [valid] pointer to a [`ffi::RSScoreExplain`] and must
+    /// **stay** valid for the entire lifetime of the returned [`SearchResult`].
+    ///
+    /// [valid]: https://doc.rust-lang.org/std/ptr/index.html#safety
+    // TODO resolve ownership (this is heap-allocated but owned by this search result??)
+    score_explain: Option<NonNull<ffi::RSScoreExplain>>,
+
+    document_metadata: Option<DocumentMetadata>,
+
+    // index result should cover what you need for highlighting,
+    // but we will add a method to duplicate index results to make
+    // them thread safe
+    // TODO resolve ownership
+    index_result: Option<NonNull<RSIndexResult<'a>>>,
```

Comment:
Should also be aligned with the iterators trait in the future (currently `Option<&RSIndexResult<'a>>`)

Reply:
Good point, this was actually more difficult to address since `RQEIterator` returns an immutable reference to the `RSIndexResult` whereas `SearchResult` needed to be able to hand out mutable access [here](https://github.com/RediSearch/RediSearch/blob/6daf604f053c7c713cdcc463e9888c0ae25a28d7/src/aggregate/functions/string.c#L35).

After running this past the Mainmatter Rust folks I refactored the relevant function in `index_result.h` to accept immutable pointers and removed the `SearchResult_GetIndexResultMut` function.